refactor `Servable` model wrappers to instantiate `from_params` instead of using default values

joelgrus opened the pull request:
and create them from persisted `Vocabulary`s instead of from datasets

I thought about doing a `Servable.from_params` and using the registry, but I didn't see an immediate benefit.

matt-gardner commented on allennlp/service/servable/models/semantic_role_labeler.py:
```diff
@@ -55,3 +39,23 @@ def predict_json(self, inputs: JSONDict) -> JSONDict:
+            token_indexers[name] = TokenIndexer.from_params(indexer_params)
+
+        vocab_dir = params.pop('vocab_dir')
+        vocab = Vocabulary.from_files(vocab_dir)
```
Shouldn't this just be the model serialization prefix?  Presumably you're going to want to load a model here, too.

I'm wondering if there's just an easier way to do this.  If you load the experiment configuration file, you can pull out the model params, instantiate the model (which already has a vocab dir), pull out the dataset reader and grab its tokenizer and token indexers, and not have to do all of the individual processing and instantiation that's done here.  That way you don't have to duplicate the configuration files or the loading code, you just pull out the pieces you need after the already-existing code runs.

And if we go that route, I don't think I'd call this method `from_params`, I'd probably call it `from_experiment_config`, or `from_experiment_params`, or something.  There's no need to have a `from_params` method for this if it doesn't make sense.  The models and things themselves should definitely be instantiated `from_params`, but that doesn't mean this class has to match the same API.

What do you think?

joelgrus replied:
I tried this on the bidaf one, take a look and see if it's what you had in mind, if so I can do the other ones too

joelgrus commented on allennlp/service/servable/servable.py:
```diff
@@ -38,16 +40,24 @@ def register(self, key: str, servable: Servable):
         import allennlp.service.servable.models.decomposable_attention as decomposable_attention
 
+        bidaf_params = Params(replace_none(pyhocon.ConfigFactory.parse_file(
+                'allennlp/service/servable/models/bidaf.conf')))
```
We need to get rid of the `/net/efs` stuff anyway since it only works on ai2 ec2 machines. I guess I should go off and think about what's the right thing to do that works well for ai2 people and non-ai2 people.

matt-gardner replied:
Yeah, but as long as we have actual paths in the configuration files that point to real data, they are going to have to be changed in a test, or by a third-party user of the code.  Unless they point to some s3 location where we've stored all of the data we refer to...  But in that case we should still probably just modify it for tests.

matt-gardner commented on tests/service/servables/bidaf_test.py:
```diff
@@ -13,27 +13,27 @@ def test_uses_named_inputs(self):
-                        },
-                        "token_characters": {
-                                "type": "characters"
+        bidaf_config = Params({
```
Can you just load the file from `experiment_config` here, too, like you do in `servable.py`?

matt-gardner commented on allennlp/service/servable/models/bidaf.py:
```diff
@@ -33,24 +34,28 @@ def predict_json(self, inputs: JSONDict) -> JSONDict:
 
         token_indexers = {}
-        token_indexer_params = params.pop('token_indexers')
+        token_indexer_params = config["dataset_reader"].pop('token_indexers')
```
I _think_ it's safe and fast to just instantiate the `DatasetReader` itself here.  Then you can pull out the `_tokenizer` and `_token_indexer` fields directly from the reader, and have 3-4 lines of code, instead of 10, for this.  It'd be reading the internal fields of the `DatasetReader`, but I don't have a huge problem with that in this case.

If you _do_ really want to keep it this way, the tokenizer params are inside "dataset_reader", not at the top level - we just use the default in BiDAF, so it's not obvious where they would have shown up.

joelgrus commented:
ok, I think this is pretty much how we discussed. a few things

(1) I had to add a fake decomposable attention spec. I assume you will replace it with a real one soon.
(2) in doing so I couldn't get the embedding part right (everything I tried led to pytorch errors because I wasn't specifying the rest of the model correctly), so I still override the decomposable attention embeddings with `constants.GLOVE_FILE`. presumably once there is a real spec I can get rid of this
(3) technically `DatasetReader` doesn't have token indexers or tokenizers, only subclasses do, so I had to use specific `Subclass.from_params` instead of `DatasetReader.from_params` (the alternative was sticking a bunch of `type: ignore` annotations in)
(4) furthermore, `SrlReader` doesn't have a tokenizer (nor does the srl spec), so I just used the default one

matt-gardner commented on allennlp/service/servable/models/bidaf.py:
```diff
@@ -39,3 +32,25 @@ def predict_json(self, inputs: JSONDict) -> JSONDict:
+
+        model_params = config.pop("model")
+        assert model_params.pop("type") == "bidaf"
+        model = BidirectionalAttentionFlow.from_params(vocab, model_params)
```
Hmm...  This seems like the wrong way to load models.  You shouldn't have to instantiate the vocab separately, and _then_ load the model from params; you should be able to just load a model from a config file, without any additional parameters (where the difference between this and what we have now is that `from_params` is how you need to construct it at _training_ time, whereas if you have an already-trained model, you shouldn't have to do it the same way).

How you've done this is correct for how the code is now; I'm just saying that we should make a better way to do this soon, both for this and for the `evaluate` command.  @DeNeutoy, you agree?

matt-gardner commented:
@DeNeutoy, on your point 1, would you prefer having two separate configuration files to just reading the same one?  I thought this made things simpler and less error-prone.

Pros of this way:
- You're serving the result of a training run.  Because it's the result of training, you really _shouldn't_ need anything other than the training configuration to load and serve the model.  If you need another configuration, it will necessarily duplicate aspects of the training config, and thus be error-prone.

Cons of this way:
- You can only serve the result of a training run.  If for some reason you want to serve something else, you have to hack together something that looks like a training config file.  Not sure _why_ you would want to serve something other than a model that we trained, though...

joelgrus commented:
I am happy with only being able to serve models that we trained. I also added in the `model.eval()` calls and removed an `InitializerApplicator()` call that I don't think belonged.
